src/cluster/mcl.cpp

```diff
@@ -0,0 +1,570 @@
+	// 		cout<< " " << item;
+	// 	}
+	// 	cout<<endl;
+	// }
+	// return;
+
+	const size_t seq_count = db->ref_header.sequences;
+	statistics.reset();
+	config.command = Config::blastp;
+	config.no_self_hits = false;
+	string format = config.cluster_similarity;
+	if(format.empty()){
+		format = "qcovhsp/100*scovhsp/100*pident/100";
+	}
+	config.output_format = {"clus", format};
+	config.algo = 0;
+	config.index_mode = 0;
+	config.freq_sd = 0;
+
+	Workflow::Search::Options opt;
+	opt.db = &(*db);
+	opt.self = true;
+	SparseMatrixStream<float> ms(db->ref_header.sequences);
+	opt.consumer = &ms;
+	opt.db_filter = nullptr;
+	Workflow::Search::run(opt);
+
+	task_timer timer;
+	timer.go("Computing independent components");
+	auto componentInfo = ms.getComponents();
+	vector<vector<uint32_t>> indices = get<0>(componentInfo);
+	vector<vector<Eigen::Triplet<float>>> components = get<1>(componentInfo);
+	vector<uint32_t> sort_order(components.size());
+	iota(sort_order.begin(), sort_order.end(), 0);
+	sort(sort_order.begin(), sort_order.end(), [&](uint32_t i, uint32_t j){return indices[i].size() > indices[j].size();});
+
+	uint32_t nComponents = count_if(indices.begin(), indices.end(), [](vector<uint32_t> v){ return v.size() > 0;});
+	uint32_t nComponentsLt1 = count_if(indices.begin(), indices.end(), [](vector<uint32_t> v){ return v.size() > 1;});
+	timer.finish();
+	cout << "Found " << nComponentsLt1 <<" ("<<nComponents<< " incl. singletons) disconnected components" << endl;
+
+	timer.go("Clustering components");
+	bool full = false;
+	// Note, we will access the clustering_result from several threads below and a vector does not guarantee thread-safety in these situations.
+	// Note also, that the use of the disjoint_set structure guarantees that each thread will access a different part of the clustering_result
+	uint32_t* clustering_result = new uint32_t[db->ref_header.sequences];
+	for(uint64_t iseq = 0; iseq < db->ref_header.sequences; iseq++){
+		clustering_result[iseq]  = numeric_limits<uint32_t>::max();
+	}
+
+	uint32_t nThreads = min(config.threads_, nComponentsLt1);
+	// note that the disconnected components are sorted by size
+	const uint32_t max_counter = sort_order.size();
+	float* max_sparsities = new float[nThreads];
+	float* min_sparsities = new float[nThreads];
+	const float inflation = (float) config.cluster_mcl_inflation;
+	const float expansion = (float) config.cluster_mcl_expansion;
+
+	atomic_uint n_clusters_found(0);
+	atomic_uint component_counter(0);
+	atomic_uint n_dense_calculations(0);
+	atomic_uint n_sparse_calculations(0);
+	atomic_uint nClustersEq1(0);
+
+	auto mcl_clustering = [&](const uint32_t iThr){
+		uint32_t n_dense = 0;
+		uint32_t n_sparse = 0;
+		uint32_t n_singletons = 0;
+		uint32_t cluster_id = iThr;
+		//uint32_t my_counter = component_counter.fetch_add(1, memory_order_relaxed);
+		//while(my_counter < max_counter){
+		for(uint32_t my_counter = iThr; my_counter<max_counter; my_counter += nThreads){
```

Comment:
The solution with `component_counter` seems to be better here.

Reply:
Yes, but the issue was that I have never observed a cpu usage over 100% even though the threads have been spawned. My assumption was that the fetch_add call was blocking (on the system I was running on - I didnt check the is_lock_free() function). But also with the changes here the cpu usage in the clustering phase was at a 100%, so I am a bit puzzled. Considering that for the test I have been running, the clustering is 100x faster than creating the adjacency matrix, this is probably not an issue (yet). But we may want to rethink whether the alignment phase is actually necessary for creating an (approximate) adjacency matrix

Reply:
I have just confirmed that is_lock_free is true. I am a bit puzzled what is going on. I will switch back to the solution with the atomic update but change the granularity, such that each of the threats takes a few iterations in one shot, eg. 100 or so (could be made dependent on the number of threats and total number of disconnected components).